### PR TITLE
Diffmode

### PR DESCRIPTION
**Pull request by pdedes**
Diffmode introduces a global setting that can be switched on and off in the scrubber. 
*It also switches us from the diff_match_patch library to jsDiff
*Asynchronicity of diff calculations has been handled by promisifying jsDiff
*A new SettingsFactory is available for saving global application settings/modes
*A bug remains that prevents you from loading a new file while diffMode is ON. This is a result of when 'broadcastKeyframe' inside of 'updatePointers' in scrubber. It's a minor issue that doesn't cause any major functionality losses.
diffmode

### CODE REVIEW

File: js/factory/errors.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+
+app.factory('Errors', function($timeout) {
```
Comment (carolynyoo) [inline js/factory/errors.js]: naming convention - conform to ErrorsFactory? (or change others to not have 'factory' in them)

File: js/factory/settingsfactory.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+	// Store booleans as private vars and expose functions for getting or setting them.
+	
+	var diffMode = false;
+
+	return {
+		getMode: function () {
+			return diffMode;
+		},
+		switchMode: function () {
+			diffMode = diffMode === true ? false : true;
```
Comment (carolynyoo) [inline js/factory/settingsfactory.js]: is it poss to return diffMode!==diffMode instead? for brevity's sake

File: js/ui-routes/scrubber/scrubber.js [hunk 3]
```diff
@@ -37,7 +46,14 @@ app.directive('scrubber', function() {
 				if ($scope.keyframeIndex === $scope.keyframes.length-1) {
 					$scope.isLastFrame = true;
 				}
-				$scope.broadcastKeyframeSelected();
+
+				if (SettingsFactory.getMode()) {
```
Comment (carolynyoo) [inline js/ui-routes/scrubber/scrubber.js:50]: yay!

File: js/ui-routes/viewer/viewer.js [hunk 4]
```diff
@@ -6,11 +6,15 @@ app.directive('viewer', function() {
 		restrict: 'E',
 		templateUrl: 'js/ui-routes/viewer/viewer.html',
 		// scope: {},
-		controller: function ($scope, CommLinkFactory, KeyframeFactory) {
+		controller: function ($scope, CommLinkFactory, SettingsFactory) {
 
+			var Range = ace.require("ace/range").Range;
```
Comment (carolynyoo) [inline js/ui-routes/viewer/viewer.js:11]: oooo nicee

File: js/factory/errors.js (hunk 1 above)
Comment (pdedes) [inline js/factory/errors.js]: udpated

File: js/factory/settingsfactory.js (hunk 2 above)
Comment (pdedes) [inline js/factory/settingsfactory.js]: I presume your'e referring to func 'switchMode'? I've shortedned it further actually, I'm going to just have it 'return diffMode = !diffMode;'

Comment (carolynyoo) [inline js/factory/settingsfactory.js]: yeeee

**Pull request merged**